### PR TITLE
Refresh cached postcodes

### PR DESCRIPTION
**Pull request by catalinailie**
Trello card: https://trello.com/c/PnzlTlbU/2795-implement-cache-warming-up-process-for-locations-api-5

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

**Review by BhupsAL1 — APPROVED** [review]
LGTM

File: spec/lib/os_places_api/client_spec.rb [hunk 1]
```diff
@@ -182,6 +182,62 @@
       end
     end
 
+    context "the postcode is outdated in the database" do
+      before do
+        Postcode.create(
+          postcode: postcode,
+          results: [
+            {
+              "DPA" => {
```
Comment (ChrisBAshton) [inline spec/lib/os_places_api/client_spec.rb:191]: It would be great to not need to duplicate all this JSON - if the JSON ever changes, we'd have to update it in multiple places. I'd suggest taking an approach like this:

```ruby
  arbitrary_value = -0.1
  old_results = os_places_api_results.dup
  old_results.first["DPA"]["LNG"] = arbitrary_value
  old_results.first["DPA"]["LAT"] = arbitrary_value
  Postcode.create(postcode: postcode, results: old_results)
```

File: spec/lib/os_places_api/client_spec.rb [hunk 2]
```diff
@@ -182,6 +182,62 @@
...
+                "LANGUAGE" => "EN",
+                "MATCH" => 1.0,
+                "MATCH_DESCRIPTION" => "EXACT",
+              },
+            },
+          ],
+        )
+      end
+
+      it "should query OS Places API and update cached data" do
```
Comment (ChrisBAshton) [inline spec/lib/os_places_api/client_spec.rb:227]:
```suggestion
      it "should query OS Places API and update cached data when `update: true` is passed" do
```

I'd also add another test with `update: false` passed, that checks that the data _isn't_ updated (or rather that no HTTP requests are made to OS Places API).

File: lib/os_places_api/client.rb [hunk 3]
```diff
@@ -25,7 +25,11 @@ def locations_for_postcode(postcode)
         json = JSON.parse(response.body)
         raise UnexpectedResponse if json["results"].nil?
 
-        Postcode.create!(postcode: postcode, results: json["results"]) unless Postcode.find_by(postcode: postcode)
+        if (existing = Postcode.find_by(postcode: postcode))
+          existing.update(results: json["results"]) if update
+        else
+          Postcode.create!(postcode: postcode, results: json["results"])
+        end
```
Comment (ChrisBAshton) [inline lib/os_places_api/client.rb:32]: Nice 👍 logic changes look good.

File: config/initializers/sidekiq.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Sidekiq.configure_server do |config|
+  # Calls to Rails.logger in a sidekiq process will use Sidekiq's logger
+  Rails.logger = Sidekiq::Logging.logger
```
Comment (ChrisBAshton) [inline config/initializers/sidekiq.rb:3]: I assume this bit of config was copied from another repo. Could you link to https://github.com/alphagov/email-alert-api/commit/529711e8d72a272c352a373951a348f179c28d4b from your commit message? That way we have an easy way of finding out why the config was added.

File: config/initializers/sidekiq.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+
+  config.on(:startup) do
+    already_scheduled = Sidekiq::JobSet.new.any? { |job| job.klass == "PostcodesCollectionWorker" }
```
Comment (ChrisBAshton) [inline config/initializers/sidekiq.rb:6]: Checking for an already scheduled worker feels like a good idea, just wondering if you came across this in testing or you're just being cautious? I'd usually advise adding in a guard like this only if it becomes necessary, i.e. just have `PostcodesCollectionWorker.perform_async` in this block.

File: app/workers/application_worker.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class ApplicationWorker
+  include Sidekiq::Worker
+
+private
+
+  def run_with_advisory_lock(model, unique_ref, &block)
```
Comment (ChrisBAshton) [inline app/workers/application_worker.rb:6]: This method is only used by the `PostcodesCollectionWorker`, so I wouldn't bother using inheritance here, I'd just move this method directly into that class, and delete this class.

File: app/workers/postcodes_collection_worker.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class PostcodesCollectionWorker < ApplicationWorker
+  sidekiq_options queue: :collect_postcodes
+
+  def perform
+    run_with_advisory_lock(Postcode, postcode) do
+      postcodes = Postcode.select(:postcode)
+
+      postcodes.each do |postcode|
+        ProcessPostcodeWorker.perform_async_in_queue(postcode, queue: :process_postcode)
```
Comment (ChrisBAshton) [inline app/workers/postcodes_collection_worker.rb:9]: I think this will have the effect of immediately putting 12 million workers on a queue (once the database has been populated with all postcodes). How quickly this queue is processed is then down to each individual worker. For example, if the `ProcessPostcodeWorker` only takes 10ms to process each postcode, we'll work through 100 workers in a second, and could very quickly hit a rate limit on OS Places API (or worse, bring it down!).

I think this is the point where we need to artificially slow things down, probably by adding a `sleep 1` after line 9, so that only 1 worker is created every second. This keeps us safely in the limits of what is accepted by OS Places API.

File: app/workers/postcodes_collection_worker.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+class PostcodesCollectionWorker < ApplicationWorker
+  sidekiq_options queue: :collect_postcodes
+
+  def perform
+    run_with_advisory_lock(Postcode, postcode) do
```
Comment (ChrisBAshton) [inline app/workers/postcodes_collection_worker.rb:5]: The `sleep` I've suggested might be problematic with the `run_with_advisory_lock` block though. Reading the [docs](https://www.rubydoc.info/gems/with_advisory_lock/3.0.0), it implies to me that the database is locked until the lock is released, which in this case would take 12 million seconds! I think we're probably better off removing the `run_with_advisory_lock` method (unless I'm missing something?).

File: app/workers/postcodes_collection_worker.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+      postcodes = Postcode.select(:postcode)
+
+      postcodes.each do |postcode|
```
Comment (ChrisBAshton) [inline app/workers/postcodes_collection_worker.rb:8]: When all the postcodes have been updated once, that's the end of this worker's work. As far as I can see, we don't yet have a mechanism for forcing another refresh. I think we'll need to wrap the contents of our `perform` method with a `while true`, i.e. when the `postcodes.each` iteration comes to an end, we'll want to run it all over again.

 We'll want a test for that too. Therefore probably worth doing in a second commit.

File: config/sidekiq.yml [hunk 10]
```diff
@@ -1,2 +1,8 @@
 ---
 :concurrency: 1
+:logfile: ./log/sidekiq.json.log
+:timeout: 4
+:max_retries: 9
```
Comment (ChrisBAshton) [inline config/sidekiq.yml:5]: I see the above two values are [taken from email-alert-api](https://github.com/alphagov/email-alert-api/blob/df501d3edb7185ce35ab5592489765ec3699dae5/config/sidekiq.yml#L5-L6).

The use case for Locations API is different - if a postcode update fails, it's less important than if an email fails to be sent. I think it's reasonable to keep retries low. The [default is 25](https://github.com/mperham/sidekiq/wiki/Error-Handling#automatic-job-retry), which is too much; I think we can actually set this to `1`.

The [default timeout is 25s](https://github.com/mperham/sidekiq/blob/152c3d8104295afc5b34ed9c190eaa85806b722f/lib/sidekiq.rb#L26), and again it seems reasonable to cut this value down. 4s seems fair 👍

File: config/sidekiq.yml [hunk 11]
```diff
@@ -1,2 +1,8 @@
...
+:queues:
+  - [collect_postcodes, 4]
+  - [process_postcode, 2]
```
Comment (ChrisBAshton) [inline config/sidekiq.yml:8]: I think we can remove these and just use `- default` - see https://github.com/mperham/sidekiq/wiki/Advanced-Options#queues

Setting queue priorities is useful when we're bulk-processing lots of different workers at the same time, but we're planning to only ever have 2 active workers: one long-lived `PostcodesCollectionWorker`, and many many short-lived `ProcessPostcodeWorker` (one at a time).

**Review by ChrisBAshton — COMMENTED** [review]
Hi @catalinailie - nice work putting this together 👍 I'm happy that it will work, but have a few comments mostly around configuration.

Nitpick: please also tweak your commit messages so that the first line is <50 chars, so that it displays succinctly on GitHub.

File: spec/lib/os_places_api/client_spec.rb [hunk 12]
```diff
@@ -182,6 +182,44 @@
...
+        expected_results.first["longitude"] = -1.0
+
+        expect(client.locations_for_postcode(postcode, update: false).as_json).to eq(
+          {
+            "average_latitude" => -1.0,
+            "average_longitude" => -1.0,
+            "results" => expected_results,
+          },
+        )
+      end
```
Comment (ChrisBAshton) [inline spec/lib/os_places_api/client_spec.rb:220]: Nice tests 👍  it feels like the addition of the `update: true` parameter can be in its own commit though, and then we can create the `ProcessPostcodeWorker` sidekiq worker in a separate commit. They're two separate changes.

**Review by ChrisBAshton — COMMENTED** [review]
@catalinailie this is looking great 👍  a couple of minor comments below - my biggest concern is that if we remove the `sleep 1` between worker instantiations, the tests would still pass, but we'd introduce a big performance problem. So would be good to introduce some sort of safeguard test for that. Not quite sure how we'd implement that yet though - would take a bit of digging.

File: spec/workers/postcodes_collection_worker_spec.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        allow(Postcode).to receive(:select).and_return(%w[E18QS E18QL])
+      end
+
+      it "creates a worker for each postcode" do
+        expect(ProcessPostcodeWorker).to receive(:perform)
+          .with("E18QS")
+        expect(ProcessPostcodeWorker).to receive(:perform)
+          .with("E18QL")
+
+        # expect(sleep).to receive(1)
```
Comment (catalinailie) [inline spec/workers/postcodes_collection_worker_spec.rb:18]: Still trying to find a way how to test that the worker is waiting 1 second before each postcode. I tried to check that the `perform` call is not taking less than 2s but it still works if I remove the sleep.

File: app/workers/postcodes_collection_worker.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
+class PostcodesCollectionWorker
+  include Sidekiq::Worker
+
+  sidekiq_options queue: :collect_postcodes
+
+  def perform(test: false)
+    loop do
+      postcodes = Postcode.select(:postcode)
```
Comment (ChrisBAshton) [inline app/workers/postcodes_collection_worker.rb:8]: Just doing a quick bit of arithmetic on this: what kind of performance issues might we hit by selecting 12 million rows? 🤔 

Assuming each postcode is about 7 characters long, therefore 7 bytes, this query might put 84 million bytes in memory, i.e. 82031 KB or 80.1 MB. Shouldn't be an issue.

But as a quick test, I spun up a Rails console in Whitehall and did a `Edition.select(:document_id)`, which [has an index](https://github.com/alphagov/whitehall/blob/5d45cad7c2606f0fc0fa29d784de2bc7c9bd8fe9/db/schema.rb#L423) like [our postcode field does](https://github.com/alphagov/locations-api/blob/e7d750118d362e519704afb8c94717473845721d/db/schema.rb#L23). The query took a few minutes and ended with `(Object doesn't support #inspect) => ` - not encouraging.

So I tried `Document.select(:id)`. Again, it took a minute or two. Ended up getting 378378 results. That's 31x fewer results than we'll have in Locations API - which does make me wonder if we might be stuck for around half an hour retrieving all the rows with this `select` query!

That's probably _not_ the case - I doubt the memory allocation is that linear - but even if it was, I don't think that would necessarily be an issue either. [Long-running jobs are a thing](https://stackoverflow.com/questions/58353386/sidekiq-jobs-maximum-run-time), so it's not like Sidekiq would crash.

TLDR: I thought we may have a performance issue on this line, but on closer inspection, I don't think we do 😄

File: spec/workers/postcodes_collection_worker_spec.rb [hunk 15]
```diff
@@ -0,0 +1,24 @@
+RSpec.describe PostcodesCollectionWorker do
+  describe "#perform" do
+    def perform
+      described_class.new.perform
+    end
+
+    context "where there are postcodes to be refreshed" do
+      before do
+        allow(Postcode).to receive(:select).and_return(%w[E18QS E18QL])
```
Comment (ChrisBAshton) [inline spec/workers/postcodes_collection_worker_spec.rb:9]: I don't think `select` returns the values - it returns a scaled back instance of `Postcode`. See equivalent example in Whitehall:

```
Document.limit(2).select(:id)
=> [#<Document:0x00007f60fa796b18 id: 4377>, #<Document:0x00007f60fa796a50 id: 4397>]
```

What you're probably looking for is `pluck`:

```
Document.limit(2).pluck(:id)
=> [4377, 4397]
```

File: spec/workers/postcodes_collection_worker_spec.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+      end
+
+      it "creates a worker for each postcode" do
+        expect(ProcessPostcodeWorker).to receive(:perform)
+          .with("E18QS")
+        expect(ProcessPostcodeWorker).to receive(:perform)
+          .with("E18QL")
+
+        expect(subject).to receive(:sleep).exactly(2).times
```
Comment (ChrisBAshton) [inline spec/workers/postcodes_collection_worker_spec.rb:18]: Nice job testing this!

File: spec/workers/postcodes_collection_worker_spec.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+        subject.perform(test: true)
```
Comment (ChrisBAshton) [inline spec/workers/postcodes_collection_worker_spec.rb:20]: I'd suggest inverting the parameter here - change it to `run_continuously: false` (but set to `true` by default).
We may legitimately want to only do single run-throughs someday, so this would give us that feature.
Rather than telling the application code "this is a test, so do something different", we're specifying in the test how we want the application code to run.
Hope that makes sense!

**Review by ChrisBAshton — COMMENTED** [review]
Thanks for making the changes, @catalinailie! This looks good to me. Just one blocking comment, which is that you may need to swap for a `pluck` to get the response you want.

Once that's addressed, we should build this branch to Integration and test that it does what we think it does (probably via some combination of [Sidekiq monitoring](https://docs.publishing.service.gov.uk/manual/sidekiq.html#sidekiq-web-aka-sidekiq-monitoring) and checking that the `Postcode.last.updated_at` value gets continually updated in the Locations API Rails console.

File: spec/workers/postcodes_collection_worker_spec.rb (hunk 15 above)
Comment (catalinailie) [reply spec/workers/postcodes_collection_worker_spec.rb:9]: Ohh I see, thanks. I wasn't aware of this 🤔

**Review by ChrisBAshton — APPROVED** [review]
Excellent work 🎉  LGTM, pending a test on Integration.

**Review by ChrisBAshton — APPROVED** [review]
Nice one 🎉
